src/base/utils/version.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016  Eugene Shalygin
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef QBITTORRENT_UTILS_VERSION_H
+#define QBITTORRENT_UTILS_VERSION_H
+
+#include <array>
+#include <stdexcept>
+
+#include <QDebug>
+#include <QString>
+#include <QStringList>
+
+namespace Utils
+{
+
+    template <class T, std::size_t N, std::size_t Mandatory = N>
+    class Version
+    {
+        static_assert(N > 0, "The number of version components may not be smaller than 1");
+        static_assert(N >= Mandatory,
+                      "The number of mandatory components may not be larger than the total number of components");
+
+    public:
+        typedef T ComponentType;
+        typedef Version<T, N, Mandatory> ThisType;
+
+        constexpr Version()
+            : m_components {}
+        {
+        }
+
+        constexpr Version(const ThisType &other) = default;
+
+        template <class ... Other>
+        constexpr Version(Other ... components)
+            : m_components {{components ...}}
+        {
+        }
+
+        Version(const QString &version)
```

Comment:
So it can throw exception, IIRC.
Please add appropriate comments to methods/constructors that can throw exception, but doesn't contain `throw` statement in its body directly.